src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -761,7 +761,8 @@ void ObjectCopyRequest<I>::compute_zero_ops() {
         if (object_extent.offset + object_extent.length >= end_size) {
           // zero interval at the object end
           if ((object_extent.offset == 0 && hide_parent) ||
-              (object_extent.offset < prev_end_size)) {
+              (object_extent.offset < prev_end_size) ||
+              (m_src_snap_id_start != 0 && prev_end_size == 0)) {
```

Comment:
> So, if for example the previous snapshot (which is not a part of this copy, so we don't know its actual `end_size`) actually has `end_size <= object_extent.offset` we will unnecessary add this region as `io::SPARSE_EXTENT_STATE_ZEROED` to `m_snapshot_sparse_bufferlist`.
> 
> On the other hand, at first glance it seems like it should not cause any problem (just one unnecessary truncate/rm op) and actually if I don't miss something all this `if` condition looks like an optimization to me, that we could just remove and unconditionally add the region to to `m_snapshot_sparse_bufferlist`. I would try this, though it would need thorough testing.

Wouldn't we still need to check hide_parent in this case?

Reply:
My idea is to add truncate operation unconditionally, so sure no need to check for hide_parent here. But note, it was long time ago when I was looking at this code and I may be missing something. I hope the tests will catch it if I am wrong. And that is why I suggest to start with improving our tests. Actually I think currently we have good enough tests that cover full image deep copy. And we just need to extend them to also test copying a subset of snapshots.